script: add script to generate example bitcoin.conf

josibake opened the pull request:
create a script for parsing the output from `bitcoind --help` to create an example conf file for new users

## problem

per #10746 , `bitcoin.conf` not being put into the data directory during installation causes some confusion for users when running bitcoin. in the discussion on the issue, one proposed solution was to have an example config file and instruct users to `cp` it into their data directory after startup. in addition to #10746 , there have been other requests for a "skeleton config file" (https://github.com/bitcoin/bitcoin/issues/19641) to help users get started with configuring bitcoind.

the main issue with an example config file is that it creates a second source of truth regarding what options are available for configuring bitcoind. this means any changes to the options (including the addition or removal of options) would have to be updated for the command line and also updated in the example file.

this PR addresses this issue by providing a script to generate an example file directly from the `bitcoind --help` on-demand by running `contrib/devtools/gen-bitcoin-conf.sh`. this solution was originally proposed on #10746 and would also solve #19641 . this guarantees any changes made to the command-line options or the command-line options help would also be reflected in the example file after compiling and running the script.

the main purpose of this script is to generate a config file to be included with releases, same as `gen-manpages.sh`. this ensures every release also includes an up-to-date, full example config file for users to edit. the script is also available for users who compile from source for generating an example config for their compiled binary.

## special considerations

this removes the `bitcoin.conf` example file from the repo as it is now generated by this script. the original example file did contain extra text related to how to use certain options but going forward all option help docs should be moved into `init.cpp`

this also edits `init.cpp` to have the option help indicate that `-conf` is not usable from the config file. this is similar to how `-includeconf` 's help indicates it cannot be used from the command line

josibake commented on share/examples/bitcoin.conf:
```diff
@@ -1,175 +1,615 @@
-#keypool=100
-
-# Maintain coinstats index used by the gettxoutsetinfo RPC (default: 0).
+#alertnotify=<cmd>
```
good call-out, added an `awk` line for this

klementtan commented:
Having a generated `bitcoin.conf` would be very useful and I would personally like this to be implemented. Tested on macOs and built cleanly. 

>https://github.com/bitcoin/bitcoin/pull/21911#issuecomment-838133285: If it is produced by our build system it should go into .gitignore.

I think it might be better to add `bitcoin.conf` to `.gitignore` since it would be generated from the build system.

josibake commented:
> I think it might be better to add `bitcoin.conf` to `.gitignore` since it would be generated from the build system.

thanks for the feedback, @klementtan , i was conflicted on whether or not to remove `bitcoin.conf` from the repo, but the link you provided seems to indicate it should be. ill add it to the .gitignore

jarolrod reviewed (commented):
This is definitely a nice-to-have

The first three commits should be squashed (634becfdcf6bda55ec0c57b2f001a54ae6abaf5a, 54235a32b5b2b871c6e9d44ede743326def91a74, cf90dc75faf20ee9daeef68939cf2b6737a59e4f) because the addition of the `generate-example-bitcoin-conf.sh` script should be atomic.

In regards to: c4e94db15266e613969d7e43690f8fa69e1e8fa2
I don't think that we should remove the `bitcoin.conf` example file. We should keep the file as part of the repo. The script can update the `bitcoin.conf` example whenever there is a PR that changes an option or introduces an option.

You should update [bitcoin-conf.md](https://github.com/bitcoin/bitcoin/blob/master/doc/bitcoin-conf.md) where appropriate to mention the inclusion of this script. This can included in the commit that will be created when you squash the first three commits.

josibake commented:
thanks, @jarolrod , will definitely update the bitcoin-conf.md

> I don't think that we should remove the `bitcoin.conf` example file.

do you have a reason you think it should be committed to the repo? #21911 indicates files generated by the build shouldn't be committed.

id also be worried about someone changing either the options or option help text and not running the script. this would cause

1) weird conflicts when running in CI
2) users modifying their local copy of the repo unexpectedly just by doing a make install
3) the example conf being out of date with the command line options

if the developer always remembers to run the script, none of the above would be problems, but im hesitant to rely on developers remembering to do a manual step.

thoughts? also tagging @laanwj for input as he was the one who made the comment about build generated files being in `.gitignore` and originally suggested the idea of generating the bitcoin.conf example from `bitcoind -h` in #10746

josibake committed:
doc: update bitcoin-conf.md

include instructions on how to run the script

jonatack commented:
> Reviewers, this pull request conflicts with the following ones:
> 
>     * [doc: repository IRC updates #22349](https://github.com/bitcoin/bitcoin/pull/22349) by jonatack

Removed the one-line change behind the conflict.

shaavan reviewed (commented):
tACK 9fab9de36db3b6aae3b2b51fa14357c2fdc90f0d
Tested successfully on Ubuntu 20.04

The PR is adding a script-based functionality to generate an example bitcoin.conf file instead of having a premade example bitcoin.conf file. This is done using a script `generate-example-bitcoin-conf.sh`, which uses some predefined text and the output of `./src/bitcoind --help` to create an example `bitcoin.conf` file.

**Test failed** while using both `make install` and `./contrib/generate-example-bitcoin-conf.sh` command from root directory. **Worked** only after giving `sudo` permission to these commands.
So tested successfully using `sudo make install` and `sudo  ./contrib/generate-example-bitcoin-conf.sh`. This happened because the `./share/examples` folder was **protected** by default and could not be modified until given `sudo` access.

I agree with the changes made by this PR because now the example `bitcoin.conf` file is very highly detailed and arranged so that understanding and changing something on the file has become very easy compared to the master’s example `bitcoin.conf` file.
Also in the current master the example `bitcoin.conf` file is fixed so we have to keep updating it, with each change that might effect working of `bitcoin.conf` file. With this PR the op is able to mitigate the example `bitcoin.conf` maintenance cost, by basically generating the file from `./src/bitcoind --help` command. Which makes sure `./src/bitcoind --help` and example `bitcoin.conf` file are always in sync. 

Though I agree with the concept I would highly recommend adding the following changes to the PR:

1. In `doc/bitcoin-conf.md` file it should be mentioned that if both ways of creating the `bitcoin.conf` example file (`make file `and  `./contrib/generate-example-bitcoin-conf.sh`) fails then the issue could be that they require `sudo` permission to edit the `./share/examples` folder.

2. The PR describes the problem that it is trying to solve as:

    >per #10746 , `bitcoin.conf` not being put into the data directory during installation causes some confusion for users when running bitcoin. more details in the issue

    But in reality, the PR is not doing anything to make it easier to put example `bitcoin.conf` file in the data directory. The aim of the PR, as far as I can see, is to improve example `bitcoin.conf` file’s flexibility wrt future changes in `./src/bitcoind --help`, and to make it easy to use.
So the PR description should be modified to display its motivations more accurately. 
<p>

josibake commented:
>     2\. So the PR description should be modified to display its motivations more accurately.

good call out, i have updated the PR description to more accurately describe the problem this PR is addressing

regarding your recommendation that the docs be updated to suggest running `sudo make install`, in general i am against recommending in docs that users run commands with sudo. id rather leave it to them to evaluate the error message and determine for themselves if running with sudo is necessary. i was also not able to reproduce the issue you are describing.

while trying to recreate your issue, i did note that the script will fail if `--prefix` is used with `./configure` and have updated the script to use `src/bitcoind` when generating the file. this change is reflected in https://github.com/bitcoin/bitcoin/pull/22235/commits/9df03cb5c5abd8f31a4e93b006d5d3dd09e52a1d

josibake commented:
> i was also not able to reproduce the issue you are describing

worth mentioning, outside of this PR i normally encounter a failure running `make install` without sudo with the following error:

```make[4]: Leaving directory '/home/josibake/bitcoin'
 /usr/bin/mkdir -p '/usr/local/lib'
 /bin/bash ../libtool --preserve-dup-deps  --mode=install /usr/bin/install -c   libbitcoinconsensus.la '/usr/local/lib'
libtool: install: /usr/bin/install -c .libs/libbitcoinconsensus.so.0.0.0 /usr/local/lib/libbitcoinconsensus.so.0.0.0
/usr/bin/install: cannot create regular file '/usr/local/lib/libbitcoinconsensus.so.0.0.0': Permission denied
```
to re-create your issue, i first configured with `./configure --without-bdb --prefix=/tmp` to get around running `sudo make install` (based on this: https://askubuntu.com/questions/424786/permission-denied-while-running-make-install). this is how i uncovered the issue with using `bitcoind` vs `src/bitcoind`.

after updating the script to use `src/bitcoind` i was able to run `make install` without errors. perhaps if you could post the exact error you were getting when running `make install`?

Rspigler commented:
Love this!
tACK 98d8482d37cb3d562b3d7f1f085d75074c3f2b6d

jarolrod commented on doc/bitcoin-conf.md:
```diff
@@ -61,4 +61,9 @@ Windows | `%APPDATA%\Bitcoin\` | `C:\Users\username\AppData\Roaming\Bitcoin\bitc
 macOS | `$HOME/Library/Application Support/Bitcoin/` | `/Users/username/Library/Application Support/Bitcoin/bitcoin.conf`
 
-You can find an example bitcoin.conf file in [share/examples/bitcoin.conf](../share/examples/bitcoin.conf).
+An example configuration file is generated by [contrib/generate-example-bitcoin-conf.sh](../contrib/generate-example-bitcoin-conf.sh) during `make install`. The output is placed in [share/examples/bitcoin.conf](../share/examples/bitcoin.conf). If you don't see this file, you can generate it by running `./contrib/generate-example-bitcoin-conf.sh` from the root directory, or you can run `make insta ...
```
You can't include the link to the example configuration file here because we're removing it from the repo under this PR. This link is invalid link if left on the repo. It is only viable locally, after the user has generated the example file.

suggestions here:
``` diff
diff --git a/doc/bitcoin-conf.md b/doc/bitcoin-conf.md
index 98517f2fc..3babf176a 100644
--- a/doc/bitcoin-conf.md
+++ b/doc/bitcoin-conf.md
@@ -61,7 +61,10 @@ Windows | `%APPDATA%\Bitcoin\` | `C:\Users\username\AppData\Roaming\Bitcoin\bitc
 Linux | `$HOME/.bitcoin/` | `/home/username/.bitcoin/bitcoin.conf`
 macOS | `$HOME/Library/Application Support/Bitcoin/` | `/Users/username/Library/Application Support/Bitcoin/bitcoin.conf`
 
-An example configuration file is generated by [contrib/generate-example-bitcoin-conf.sh](../contrib/generate-example-bitcoin-conf.sh) during `make install`. The output is placed in [share/examples/bitcoin.conf](../share/examples/bitcoin.conf). If you don't see this file, you can generate it by running `./contrib/generate-example-bitcoin-conf.sh` from the root directory, or you can run `make install` again. To use, copy the example file into the correct data directory for your system and edit it there, like so:
+An example configuration file can be generated by [contrib/generate-example-bitcoin-conf.sh](../contrib/generate-example-bitcoin-conf.sh).
+You can run this script directly or have it automatically executed during the `make install` stage.
+The output is placed under `share/examples/bitcoin.conf`.
+To use the generated configuration file, copy the example file into your data directory and edit it there, like so:
```

josibake replied:
ah, good catch!

jarolrod reviewed (commented):
Approach ACK

josibake commented on contrib/generate-example-bitcoin-conf.sh:
```diff
@@ -0,0 +1,53 @@
+
+# create the footer text
+cat >> ${EXAMPLE_CONF_FILE} << 'EOF'
+# [Sections]
```
whoops, i forget why i took that out? i think i was copying this section from the lopp config generator

josibake commented on contrib/generate-example-bitcoin-conf.sh:
```diff
@@ -0,0 +1,53 @@
+
+EOF
+
+# parse the output from bitcoind --help
```
agree, the groups do get lost in the text. i tried a few options and 

```bash
### Group Name


# help text
#option=1
```

seemed to stand out well without being too cluttered. open to other suggestions

jarolrod reviewed (commented):
ACK d6eac4f0d237ba502aca8a4f6af239e369d6e9af

What i like about this is that it severely reduces any maintainance cost behind having an example `bitcoin.conf` file. Addditionally, it's nice that the generated example file will only contain relevant options. For example, if i configure without wallet functionality, then the generated `bitcoin.conf` will not include these options.

The presentation of sections in the output file looks good, thanks for picking up that suggestion!

shaavan reviewed (commented):
> I tested that this does fix the permissions issue by running the following commands:
> 
>     1. deleting the exambles directory
> 
>     2. running `sudo make install`
> 
>     3. verify directory and file created with correct permissions (local user)
> 
>     4. ran `make install` and verified no failures

I followed the steps you suggested here; I could not use `make install` without root permission even after that. I checked the created example **bitcoin.conf** permission, and I can see that it is `-rw-r--r--` 
I am adding the error I am getting while using the `make install` command without sudo:

```
 /usr/bin/mkdir -p '/usr/local/lib'
 /bin/bash ../libtool --preserve-dup-deps  --mode=install /usr/bin/install -c   libbitcoinconsensus.la '/usr/local/lib'
libtool: install: /usr/bin/install -c .libs/libbitcoinconsensus.so.0.0.0 /usr/local/lib/libbitcoinconsensus.so.0.0.0
/usr/bin/install: cannot remove '/usr/local/lib/libbitcoinconsensus.so.0.0.0': Permission denied

```

Though the error seems unrelated to this PR, I am can still not understand why sudo `make install` is creating the example bitcoin file with  `-rw-r--r--` permission.
<p>

josibake commented:
@shaavan the error you are getting is indeed unrelated to the PR and expected. this is because of permissions on `/usr/local/lib`. you can verify this by trying to do a fresh build on master and run `make install` (make sure to run `sudo make uninstall` first). you should see the same error.

to get around this error, you can set a prefix like `./configure --prefix=/tmp/test`.

as for the permissions, it is now (correctly) being created with `-rw-r--r--` because file creation and file permissions are determined in `Makefile.am`. this ensures that the correct owner and permissions are assigned regardless of whether sudo is used when running make install

shaavan reviewed (approved):
>you can verify this by trying to do a fresh build on master and run make install (make sure to run sudo make uninstall first). you should see the same error.
to get around this error, you can set a prefix like ./configure --prefix=/tmp/test.

I followed the steps you suggested and was able to successfully use the command `make install` without any problems with permission denied. Finally!

This PR would be a great addition to the codebase, potentially reducing a lot of maintenance burden. Thanks for this fantastic work, @josibake!

ACK b915efffa1d77497e1056cddd945632f414d3cad

laanwj commented:
Concept ACK.

However I don't think we should have scripts that call the built binaries in the build system. This gives problems with cross-compilation, for example. At most this can be a script that has to be manually invoked in the release process like the `gen-manpages.sh`.

Rspigler commented:
I'm getting these errors now:

> Generating example bitcoin.conf file in share/examples/
generate-example-bitcoin-conf.sh: line 16: share/examples/bitcoin.conf: No such file or directory
generate-example-bitcoin-conf.sh: line 29: src/bitcoind: No such file or directory
generate-example-bitcoin-conf.sh: line 34: share/examples/bitcoin.conf: No such file or directory
generate-example-bitcoin-conf.sh: line 37: share/examples/bitcoin.conf: No such file or directory

I'm now using `sudo make install`, not sure if that's the issue.

Rspigler commented:
Should bitcoin-conf.md be expanded as well? (could be done in another PR)

josibake commented:
> However I don't think we should have scripts that call the built binaries in the build system.

agreed, I ended up coming to the same conclusion trying to get this to work. I'll rework this to just be a script that can be called by the user and add a note to `release-process.md`

josibake committed:
build: include bitcoin.conf in build outputs

copy over bitcoin.conf during the build process.
this means `contrib/devtools/gen-bitcoin-conf.sh` will need
to be run and the generated file committed during the release process.

this is the same process used for generating man pages for each release.

josibake committed:
script: add script to generate example bitcoin.conf

this ensures bitcoind option help is the source of truth and also
gives an example conf file for users to customize and copy to their
data directory.

closes #10746

josibake committed:
doc: update devtools, release-process readmes

include running `gen-bitcoin-conf.sh` as part of the release process.

josibake commented:
> generate-example-bitcoin-conf.sh: line 29: src/bitcoind: No such file or directory

@Rspigler I'm not sure about the other errors, but this one is likely because there is no `bitcoind` binary present. I've added a check to the script to make sure the binary exists.

> Should bitcoin-conf.md be expanded as well? (could be done in another PR)

I'm not sure what you mean by "expanded." For now, I've included directions on how to run the script. If you're referring to including more examples on how to use `bitcoin.conf`, then yes, I think it could be expanded but in a later PR.

josibake commented:
UPDATE: I have made some minor edits to the script (inspired from `gen-manpages.sh`) and taken it out of `make install`. Users will need to run the script locally to generate a file.

I've also included generating this file as part of the release process, same as `gen-manpages.sh`. This entails running the script and committing the file before building release candidates. If the file exists, it will be copied over into the build outputs.

Rspigler commented:
tACK 0f7f46d3b590c19f3510a81a9a4648bfed63e85a works great for me now! Really love this.

w0xlt reviewed (approved):
tACK 0f7f46d
`contrib/devtools/gen-bitcoin-conf.sh` was run successfully and `bitcoin.conf` was created in `share/examples`.
Very useful script.

theStack reviewed (commented):
Conecpt ACK, very neat idea! 👌 

Nit that won't affect most users, but is probably still worth mentioning: I tested executing the script on OpenBSD 7.0 and it seeems that each option category headline has an unexpected character 'n' on the start of the line:
```
.....
n### Connection options
.....
n### Wallet options
.....
n### ZeroMQ notification options
.....
n### Debugging/Testing options
.....
n### Chain selection options
.....
n### Node relay options
.....
n### Block creation options
.....
n### RPC server options
.....
```
Didn't investigate deeper what's the cause for this, but obviously one of the used tools `awk`/`sed` behave slightly differently than on Linux. Portability issues like this are one of the reasons why we usually prefer python instead of bash scripts (see https://github.com/bitcoin/bitcoin/issues/18132 as an example).
On the other hand re-implementing short sed/awk-scripts in python could be quite a burden and lead to bloated code, so I think it's fine as it is.
If this specific issue can't be resolved, maybe check the `uname -s` output and abort if it is not `Linux`? I'd prefer restricting to certain OSes (in case of Linux, a very popular one) rather than generating a bitcoin.conf that doesn't work.

josibake commented:
> obviously one of the used tools `awk`/`sed` behave slightly differently than on Linux

thanks for catching this @theStack ! it is indeed due to BSD sed not recognizing `\n` as a newline character. I found this as a solution: https://stackoverflow.com/a/24575385

you can check with `git range-diff  0f7f46d...4bb3cf3`

can you verify this fixes the issue for you? there is also a newline character used in the `awk` command, but i don't think this causes an issue for BSD.

if this doesn't fix it, one other option is to remove the newline as it is only there for cosmetic reasons (to make the different sections more obviously differentiated). im also not opposed to re-writing it in python, but agree that feels a bit heavy for this use-case

theStack commented on contrib/devtools/gen-bitcoin-conf.sh:
```diff
@@ -0,0 +1,79 @@
+EOF
+
+# parse the output from bitcoind --help
+# line 54 is a bit funky to ensure portability for BSD
```
Referring to concrete line numbers in comments is always a bit fragile, since it's very likely forgotten to be updated if e.g. unrelated lines above are inserted. Maybe "line-break in next-to-last sed" or alike?

josibake replied:
done!

theStack reviewed (commented):
> ...
> can you verify this fixes the issue for you? there is also a newline character used in the awk command, but i don't think this causes an issue for BSD.

Thanks for fixing, I can confirm that the latest version (top-level commit 4bb3cf3dd4182b08da8fd7ef54dba2621598b315) now also works in OpenBSD 7.0! 🎉

Running a diff on my example.conf output and the one you provided me only showed differences related to different hardware and build options (different range in number of script verification threads due to having less CPU cores, no "sandbox" option due to not running Linux, no wallet related options due to having configured `--without-wallet`), i.e. obviously everything works as expected.

One concern that arose while looking at the manually created `bitcoin.conf` file in master that is removed in this PR: there is some parts that seem to contain more information than just the one provided from the help. E.g. we have a ["Quick primer on addnode vs connect"](https://github.com/bitcoin/bitcoin/blob/d3582f2d3bfebb32316aa3974f6f27db20a610f5/share/examples/bitcoin.conf#L30-L50), also the [section on rpcauth](https://github.com/bitcoin/bitcoin/blob/d3582f2d3bfebb32316aa3974f6f27db20a610f5/share/examples/bitcoin.conf#L97-L111) is more detailled than from the help, explaining how it can be generated via the `rpcauth.py` script. I would assume it's worth it to keep this information somehow?

Also left two other nits below.

prayank23 reviewed (changes requested):
Approach NACK

Reasons:

1. Bash scripts will not be helful for Windows users and Windows is used by [75%](https://gs.statcounter.com/os-market-share/desktop/worldwide) desktop computers.

<details>
  <summary>Screenshots for Windows new Bitcoin Core setup</summary>

![image](https://user-images.githubusercontent.com/13405205/147531067-b0611d9c-952b-4cc5-a605-88ddaa5ddfe9.png)

![image](https://user-images.githubusercontent.com/13405205/147531091-94f8f687-c570-4efd-92b7-c6bed14863e3.png)

![image](https://user-images.githubusercontent.com/13405205/147531137-1fef4c0c-ab7b-4911-ae96-70329295ab2f.png)

![image](https://user-images.githubusercontent.com/13405205/147531307-66780e3d-177c-48f2-bffa-1fafd4bbe940.png)
  
![image](https://user-images.githubusercontent.com/13405205/147531729-a55bd992-8d91-4647-9bdc-abfd2b2bc07b.png)
  
![image](https://user-images.githubusercontent.com/13405205/147531842-e2bc9f85-27d0-4e81-9fde-5f84c7d4270d.png)


</details>

A better approach would be to create _bitcoin.conf_ with an option in GUI either during the first time user runs Bitcoin Core and data directory is selected or user can always create _bitcoin.conf_ from options if required however UX can be improved with some templates.

2. Open Source [Bitcoin Core Config generator](https://jlopp.github.io/bitcoin-core-config-generator/) maintained by Jameson Lopp has better UI/UX to solve this problem which can be helpful for users looking to create config files with some templates, with each option categorized and explained in a better way.

3. I don't think [contrib/devtools](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools) is the right place to add such scripts if this is an attempt to help users.

4. https://github.com/bitcoin/bitcoin/pull/13761 could be a better approach for _bitcoind_ users IMO with few changes based on review comments.

josibake commented on contrib/guix/libexec/build.sh:
```diff
@@ -387,6 +387,14 @@ mkdir -p "$DISTSRC"
                 ;;
         esac
 
+	# If a bitcoin.conf file has been generated, copy it over
+	if [ -f "${DISTSRC}/share/examples/bitcoin.conf" ]
+	then
+		cp "${DISTSRC}/share/examples/bitcoin.conf" "${DISTNAME}/"
+	else
+		echo "${DISTSRC}/share/examples/bitcoin.conf file not found. Make sure to run contrib/devtools/gen-bitcoin-conf.sh before building."
+	fi
```
good catch, i really need to fix my vim setup :roll_eyes:

josibake commented:
thanks for the review, @theStack - I've incorporated your feedback here `git range-diff  4bb3cf3...bc84a32` 

> One concern that arose while looking at the manually created `bitcoin.conf` file in master that is removed in this PR: there is some parts that seem to contain more information than just the one provided from the help. E.g. we have a ["Quick primer on addnode vs connect"](https://github.com/bitcoin/bitcoin/blob/d3582f2d3bfebb32316aa3974f6f27db20a610f5/share/examples/bitcoin.conf#L30-L50), also the [section on rpcauth](https://github.com/bitcoin/bitcoin/blob/d3582f2d3bfebb32316aa3974f6f27db20a610f5/share/examples/bitcoin.conf#L97-L111) is more detailled than from the help, explaining how it can be generated via the `rpcauth.py` script. I would assume it's worth it to keep this information somehow?

personally, im not a fan of that information living in the comments of the config file. in the pr description, i mention that all help text should be moved to `init.cpp`; this ensures it will be in the generated config file and also the man pages. 

for stuff that doesn't make sense to put in `init.cpp` but that we do want to keep, i think it should go in `docs/bitcoin-conf.md`. im working on a follow-up pr to this one to add some more info to `docs/bitcoin-conf.md` and was planning to see if it made sense to move the addnode and rpcauth sections over.

josibake commented:
thanks for the feedback, @prayank23!

> Bash scripts will not be helpful for Windows users and Windows is used by 75% desktop computers.

I've included the script as part of the release process so that an up-to-date example config is included with the Windows release. For Windows users that compile from source, I believe most do it in a Linux VM/Windows subsystem for Linux/git bash/etc, so they would still be able to use the script. Check out https://medium.com/@HussainZaidi14/compiling-bitcoin-core-in-windows-10-environment-4723255e88a7 for more info.

> A better approach would be to create _bitcoin.conf_ with an option in GUI

I'm not sure what options currently exist in the GUI for editing config parameters, but I think editing/generating the config via the GUI solves a different problem and is best reserved for a separate PR.

> 2\. Open Source [Bitcoin Core Config generator](https://jlopp.github.io/bitcoin-core-config-generator/)

I feel that having an up-to-date example `bitcoin.conf` for users is important enough that it should be managed by this repository and not rely on a third party

> I don't think contrib/devtools is the right place to add such scripts if this is an attempt to help users.

The main benefit to end users is having the example `bitcoin.conf` file included with the release binaries. I chose `contrib/devtools` as the script is intended to be used during the release process and I think it keeps things simpler to have `gen-manpages.sh` and `gen-bitcoin-conf.sh` in the same place

> 4\. #13761 could be a better approach for _bitcoind_

I didn't get the sense there was strong support for the approach in #13761 and I tend to agree with the feedback here: https://github.com/bitcoin/bitcoin/pull/13761#issuecomment-408830553

In the discussion following #13761 in #10746 , I felt there was more support for this approach: https://github.com/bitcoin/bitcoin/issues/10746#issuecomment-794877769 , which was the inspiration for this PR

theStack commented:
Tested ACK bc84a32e9b8038fa59a322057b563b3f4e630919 

Works like a charm, tested config file generation on both OpenBSD 7.0 and Debian bookworm/sid by now. Didn't look deeper into the sed/awk-magic extracting and transforming the relevant information.

> personally, im not a fan of that information living in the comments of the config file. in the pr description, i mention that all help text should be moved to `init.cpp`; this ensures it will be in the generated config file and also the man pages.

Agree that the config file is not the ideal place for extended information and primers. By the way, the PR description part you mentioned (that I overlooked before) also says that `init.cpp` is changed, but it isn't currently. Was this forgotten to commit or is it not relevant anymore? In the latter case, it would probably make sense to update the PR description accordingly.

> for stuff that doesn't make sense to put in `init.cpp` but that we do want to keep, i think it should go in `docs/bitcoin-conf.md`. im working on a follow-up pr to this one to add some more info to `docs/bitcoin-conf.md` and was planning to see if it made sense to move the addnode and rpcauth sections over.

Sounds like a plan, feel free to ping me up on the follow-up PR.

prayank23 commented:
> For Windows users that compile from source, I believe most do it in a Linux VM/Windows subsystem for Linux/git bash/etc, so they would still be able to use the script. Check out https://medium.com/@HussainZaidi14/compiling-bitcoin-core-in-windows-10-environment-4723255e88a7 for more info.

Windows users that compile from source is a rare species. Also docs for compiling are available in Core repository and I have tried it: https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md

> I feel that having an up-to-date example bitcoin.conf for users is important enough that it should be managed by this repository and not rely on a third party

Example _bitcoin.conf_ already exists and there are lot of features which are not included in Core because someone wrote an open source python script which nobody uses.

> I'm not sure what options currently exist in the GUI for editing config parameters, but I think editing/generating the config via the GUI solves a different problem and is best reserved for a separate PR.

Bash script in `contrib/devs` solves problem for less than 1% users who can already find better solutions on stackexchange etc. So the OP is misleading and I do not agree with the approach.

>  I chose contrib/devtools as the script is intended to be used during the release process and I think it keeps things simpler to have gen-manpages.sh and gen-bitcoin-conf.sh in the same place

I am not sure how many users will check `contrib/devtools` for hidden scripts that help them generate config file.

> I didn't get the sense there was strong support for the approach in #13761

None of the disagreement make sense. Config file can already be generated using `bitcoin-qt`, there is nothing wrong in providing few templates.

I support the approach in #13761 and NACK on adding bash scripts that won't be helpful for most of the users.

josibake committed:
doc: update init.cpp -conf help text

update help to reflect this option cannot be used from the config file

josibake commented:
> Was this forgotten to commit

indeed, looks like i dropped this commit when re-organizing the PR! thanks for catching. I've pushed a new commit (https://github.com/bitcoin/bitcoin/pull/22235/commits/f385d175c6d566ddea11df256e45e30d9bfa6604) with the `init.cpp` edit

josibake commented:
@prayank23 there seems to be some confusion as to the intended purpose of this script. it is not intended for individual users to use for generating a file (tho it is available for use for users who are compiling from source). 

it is intended to be used as part of the release process. my apologies if the PR description confusing. I've updated the description to hopefully make this more clear:

> the main purpose of this script is to generate a config file to be included with releases, same as gen-manpages.sh. this ensures every release also includes an up-to-date, full example config file for users to edit. the script is also available for users who compile from source for generating an example config for their compiled binary.

theStack reviewed (approved):
re-ACK f385d175c6d566ddea11df256e45e30d9bfa6604

prayank23 commented:
> @prayank23 there seems to be some confusion as to the intended purpose of this script. it is not intended for individual users to use for generating a file (tho it is available for use for users who are compiling from source).
> 
> it is intended to be used as part of the release process. my apologies if the PR description confusing. I've updated the description to hopefully make this more clear:

It does not fix https://github.com/bitcoin/bitcoin/issues/10746 or https://github.com/bitcoin/bitcoin/issues/19641

Rspigler commented on doc/bitcoin-conf.md:
````diff
@@ -61,4 +61,12 @@ Windows | `%APPDATA%\Bitcoin\` | `C:\Users\username\AppData\Roaming\Bitcoin\bitc
+```
+# example copy command for linux user
+cp share/examples/bitcoin.conf ~/.bitcoin
+```
````
Add to line 74:

`Users should never make config changes which they do not understand. Furthermore, users should always be wary of accepting any config changes provided to them by another source.`

I believe this is a good compromise and would touch on @prayank23 's concerns

josibake replied:
hey @Rspigler , definitely agree that we can make some improvements to the documentation of `bitcoin-conf.md` related to security and usability, but I think that's out of scope for this PR which is solely focused on generating the conf file. Happy to address this in a follow-up!

Rspigler replied:
Agreed!

niVelion commented on contrib/devtools/README.md:
```diff
@@ -90,6 +90,20 @@ example:
+release process to include a bitcoin.conf with the release binaries and can also be run by users to generate a file locally.
+
+With in-tree builds this tool can be run from any directory within the
+repostitory. To use this tool with out-of-tree builds set `BUILDDIR`. For
```
Nit.

```suggestion
repository. To use this tool with out-of-tree builds set `BUILDDIR`. For
```

josibake replied:
thanks for catching this!

niVelion commented:
Approach ACK.

It may be worth adding some text to the top of the file to remind the user that this is an automatically generated file, how to re-generate it, and perhaps to recommend they don't edit the file "in-place" in case the script subsequently overwrites their changes. Or if that's too much documentation within the file, then some mention of where such documentation may be found.

niVelion commented on contrib/devtools/gen-bitcoin-conf.sh:
```diff
@@ -0,0 +1,79 @@
+# create the header text
+cat > "${EXAMPLE_CONF_FILE}" << 'EOF'
+##
+## bitcoin.conf configuration file. Lines beginning with # are comments.
```
```suggestion
## bitcoin.conf configuration file.
##
## This file was generated by contrib/devtools/gen-bitcoin-conf.sh.
## Lines beginning with # are comments.
```

I more commonly than not see automatically generated files such as this including some form of notice that they were generated automatically, may be worth including something like this.

josibake commented:
thanks for the review @niVelion ! ive taken both your suggestions and one small edit of my own to improve the readability of the header. changes can be verified with `git range-diff  f385d17...caf4ab7`

josibake commented:
force pushed to fix trailing whitespace :man_facepalming:

niVelion commented on contrib/devtools/README.md:
```diff
@@ -90,6 +90,20 @@ example:
+gen-bitcoin-conf.sh
+===================
+
+Generates a bitcoin.conf file in `share/examples/` by parsing the output from `bitcoind -help`. This script is run during the
```
Nit. I realize --help and -help does the same thing with the current version of bitcoind..

```suggestion
Generates a bitcoin.conf file in `share/examples/` by parsing the output from `bitcoind --help`. This script is run during the
```

josibake replied:
done!

niVelion commented on contrib/devtools/gen-bitcoin-conf.sh:
```diff
@@ -0,0 +1,82 @@
+# parse the output from bitcoind --help
+# adding newlines is a bit funky to ensure portability for BSD
+# see here for more details: https://stackoverflow.com/a/24575385
+${BITCOIND} --help | sed '1,/Print this help message and exit/d' \
```
Nit. On re-visiting, I think this reads a little better.

```suggestion
${BITCOIND} --help \
    | sed '1,/Print this help message and exit/d' \
```

josibake replied:
done!

josibake commented:
thanks for the additional suggestions @niVelion, took both suggestions for the script and readme. changes can be verified with `git range-diff  3b0ef90...120fbba`

niVelion commented:
Off-topic - entirely regarding `git range-diff`:

> thanks for the additional suggestions @niVelion, took both suggestions for the script and readme. changes can be verified with `git range-diff 3b0ef90...120fbba`

That command returns an error for me:

`$ git range-diff 3b0ef90...120fbba
fatal: ambiguous argument '120fbba..3b0ef90': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
error: could not parse log for '120fbba..3b0ef90'`

Could I ask what command sequence could I run to get this working, assuming, would be appreciated :+1: 

```
$ git remote -v
origin  git@github.com:bitcoin/bitcoin.git (fetch)
origin  git@github.com:bitcoin/bitcoin.git (push)
$ git branch
...
* master
* $ git pull
* Already up-to-date.
```

josibake commented:
> Off-topic - entirely regarding `git range-diff`

I'm not a `git` wizard (yet :smile: ) but i believe this is somehow related to you pulling down the branch after the force push? you can also verify the differences using this link: https://github.com/josibake/bitcoin/compare/3b0ef90..120fbba

MarcoFalke commented:
You can't use `...` in range-diff. I usually replace all dots between git hashes with space.

josibake commented:
> You can't use `...` in range-diff. I usually replace all dots between git hashes with space.

perhaps we are on different versions of git @MarcoFalke ? the dots are valid in range-diff (on `git 2.30.2`): 
![image](https://user-images.githubusercontent.com/7444140/149383311-e59e0821-9797-478c-9a52-58a0247023ca.png)

MarcoFalke commented:
Yes, as you can see in your screenshot: `<base> <old> <new>` doesn't use dots, but spaces.

josibake commented:
> Edit: I mean you _can_ use dots here, but then it will also print all commits that were pushed to master in the meantime.

ah, gotcha! makes sense. the spaces syntax is much more readable. TIL :smile:

MarcoFalke commented:
And for fetching commits you can copy the full sha and do `git fetch origin 120fbbafdc820f7202d9c8777a933325692f8c51`

niVelion commented:
> > Off-topic - entirely regarding `git range-diff`
> 
> @niVelion I'm not a `git` wizard (yet smile ) but i believe this is somehow related to you pulling down the branch after the force push? you can also verify the differences using this link: https://github.com/josibake/bitcoin/compare/3b0ef90..120fbba

Thanks, I tried https://github.com/bitcoin/bitcoin/compare/3b0ef90..120fbba before but not your fork.

Not certain which particular thing I tried fixed it (amongst them, adding your repo as an additional remote) but it's now working :grin:

```
$ git range-diff 3b0ef90...120fbba
1:  dbaa3e5498 ! 1:  53826de216 script: add script to generate example bitcoin.conf
    @@ contrib/devtools/gen-bitcoin-conf.sh (new)
     +# parse the output from bitcoind --help
     +# adding newlines is a bit funky to ensure portability for BSD
     +# see here for more details: https://stackoverflow.com/a/24575385
...
```

mzumsande reviewed (commented):
Concept ACK

josibake commented:
force pushed rebase; `git range-diff master 120fbba 0488dd6`

BradleyC commented:
@josibake 's point from the initial comment is extremely valid:
```
 the main issue with an example config file is that it creates a second source of truth regarding what options are available for configuring bitcoind
 ```
 as a developer new to bitcoin, it's a bit difficult to find all available / necessary options for setup on the first run. Stack and other help articles out of date, etc. 
 
 this PR solves that problem elegantly.

josibake commented:
force pushed to fix whitespace linter; `git range-diff master 0488dd6 fa67840`

willcl-ark reviewed (approved):
Partial tACK https://github.com/bitcoin/bitcoin/commit/fa678403bdc532180a435368e74b497c5dae0873

Generating the config file worked perfectly and the documentation changes read well to me.
I did not try with an out-of-tree build nor exercising `contrib/guix/libexec/build.sh`

laanwj commented on contrib/devtools/gen-bitcoin-conf.sh:
```diff
@@ -0,0 +1,83 @@
+
+export LC_ALL=C
+TOPDIR=${TOPDIR:-$(git rev-parse --show-toplevel)}
+BUILDDIR=${BUILDIR:-$TOPDIR}
```
s/BUILDIR/BUILDDIR/

laanwj commented on contrib/devtools/gen-bitcoin-conf.sh:
```diff
@@ -0,0 +1,83 @@
+BINDIR=${BINDIR:-$BUILDDIR/src}
+BITCOIND=${BITCOIND:-$BINDIR/bitcoind}
+SHARE_EXAMPLES_DIR=${SHARE_EXAMPLES_DIR:-$BUILDDIR/share/examples}
+EXAMPLE_CONF_FILE=${EXAMPLE_CONF_FILE:-$SHARE_EXAMPLES_DIR/bitcoin.conf}
```
The `gen-manpages.sh` script generates its output in the source directory, not the build directory.
This is slightly easier if the result is to be committed to the branch.

josibake replied:
good call out, this file does need to be committed

laanwj commented on doc/release-process.md:
```diff
@@ -8,6 +8,7 @@ Release Process
 * Update translations see [translation_process.md](https://github.com/bitcoin/bitcoin/blob/master/doc/translation_process.md#synchronising-translations).
 * Update release candidate version in `configure.ac` (`CLIENT_VERSION_RC`).
 * Update manpages (after rebuilding the binaries), see [gen-manpages.py](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#gen-manpagespy).
+* Generate a bitcoin.conf file, see [gen-bitcoin-conf.sh](https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#gen-bitcoin-confsh).
```
I would guess it also needs to be added to the repository, and committed?

josibake replied:
it does need to be committed for the build script to be aware of it. I went ahead and added a placeholder file (same as `doc/man/` and added a note in the docs to make it clear that the file does need to be committed. as a bonus, having a placeholder file should make it easier for users to discover the script

laanwj replied:
Sounds like a good solution to me, thanks!

laanwj commented:
Tested the script (except the guix part) and it seems to work great, left a few small nits.

josibake committed:
doc: replace bitcoin.conf with placeholder file

josibake commented:
> Tested the script (except the guix part) and it seems to work great, left a few small nits.

thanks for the review @laanwj, I took your suggestions and also added an empty placeholder file, `share/examples/bitcoin.conf` to make it more obvious that the file needs to be committed before doing a guix build

laanwj commented on contrib/guix/libexec/build.sh:
```diff
@@ -368,6 +368,14 @@ mkdir -p "$DISTSRC"
         esac
 
+        # If a bitcoin.conf file has been generated, copy it over
+        if [ -f "${DISTSRC}/share/examples/bitcoin.conf" ]
```
Now that the placeholder is always there, this if/else is no longer necessary and can be replaced with just a copy.

josibake replied:
fixed!

laanwj commented:
This seems ready for merge, sorry for finding a last-minute nit :smile:

josibake commented:
> This seems ready for merge, sorry for finding a last-minute nit smile

thanks for circling back! I've added your suggestion

laanwj commented:
Tested and code review ACK b42643c2537ffbe99d6d94fb1fb3b7f9d5234f93

hebasto commented on contrib/guix/libexec/build.sh:
```diff
@@ -368,6 +368,10 @@ mkdir -p "$DISTSRC"
 
+        # copy over the example bitcoin.conf file. if contrib/devtools/gen-bitcoin-conf.sh
+        # has not been run before buildling, this file will be a stub
+        cp "${DISTSRC}/share/examples/bitcoin.conf" "${DISTNAME}/"
```
Install a `bitcoin.conf"` file using Windows installer as well?

fanquake replied:
I'll combine this into https://github.com/bitcoin/bitcoin/pull/25829.
